File: Tuning/sample_code/01_timer/src/fortran/main.f90
! Copyright 2024 Research Organization for Information Science and Technology
!-----------------------------------------------------------------------
!  Hand-coded timer and Use of gprof 
!  Author:      Yukihiro Ota (yota@rist.or.jp)
!-----------------------------------------------------------------------
program main

  use iso_fortran_env, only: int64, real64

  implicit none

  integer,parameter :: nn = 20000
  integer :: i, k

  real(real64) :: a(nn)

#ifdef USE_ELP_TIMER
  integer(int64) :: ielp1, ielp2, icount_rate, icount_max
  real(real64) :: elp
#endif
#ifdef USE_CPU_TIMER
  real(real64) :: cpu1, cpu2
#endif

! set values
  do i = 1, nn
    a(i) = 0.0_real64
  enddo

! routine 1
#ifdef USE_ELP_TIMER
  call system_clock ( ielp1, icount_rate, icount_max )
#endif
#ifdef USE_CPU_TIMER
  call cpu_time ( cpu1 )
#endif
  do k = 1, 100000
    call sub1 ( a, nn ) 
  enddo
#ifdef USE_ELP_TIMER
  call system_clock ( ielp2, icount_rate, icount_max )
  if ( icount_rate > 0 ) then
    if ( ielp1 <= ielp2 ) then
      elp = (ielp2 - ielp1) / real(icount_rate, real64)
    else
      elp = (ielp2 - ielp1 + icount_max + 1) / real(icount_rate, real64)
    endif
    write (6,'(1a30,1f13.6)') "sub1: Elapsed time (sec)     ", elp
  else
    write (6,'(1a37)') "sub1: Elapsed time: clock unavailable"
  endif
#endif
#ifdef USE_CPU_TIMER
  call cpu_time ( cpu2 )
  write (6,'(1a30,1f13.6)') "sub1: CPU time (sec)         ", cpu2 - cpu1 
#endif
  write (6,'(1a24,1f13.6)') "a(1) =                 ", a(1)

! routine 2
#ifdef USE_ELP_TIMER
  call system_clock ( ielp1, icount_rate, icount_max )
#endif
#ifdef USE_CPU_TIMER
  call cpu_time ( cpu1 )
#endif
  do k = 1, 200000
    call sub2 ( a, nn ) 
  enddo
#ifdef USE_ELP_TIMER
  call system_clock ( ielp2, icount_rate, icount_max )
  if ( icount_rate > 0 ) then
    if ( ielp1 <= ielp2 ) then
      elp = (ielp2 - ielp1) / real(icount_rate, real64)
    else
      elp = (ielp2 - ielp1 + icount_max + 1) / real(icount_rate, real64)
    endif
    write (6,'(1a30,1f13.6)') "sub2: Elapsed time (sec)     ", elp
  else
    write (6,'(1a37)') "sub2: Elapsed time: clock unavailable"
  endif
#endif
#ifdef USE_CPU_TIMER
  call cpu_time ( cpu2 )
  write (6,'(1a30,1f13.6)') "sub2: CPU time (sec)         ", cpu2 - cpu1 
#endif
  write (6,'(1a24,1f13.6)') "a(1) =                 ", a(1)

! finalization
  stop
end
!-----------------------------------------------------------------------
!   sub 1
!-----------------------------------------------------------------------
subroutine sub1 ( a, nn )
   use iso_fortran_env, only: real64
   implicit none
   integer,intent(in) :: nn
   real(real64),intent(inout) :: a(nn)
   !*** local variables ***
   integer :: i, k

   do i = 1, nn
     a(i) = a(i) + 1.0_real64
   enddo

   do k = 1, 2
     call sub3 ( a, nn )
   enddo

   return
end subroutine 
!-----------------------------------------------------------------------
!   sub 2
!-----------------------------------------------------------------------
subroutine sub2 ( a, nn )
   use iso_fortran_env, only: real64
   implicit none
   integer,intent(in) :: nn
   real(real64),intent(inout) :: a(nn)
   !*** local variables ***
   integer :: i, k

   a(1) = sin(a(1))

   do i = 1, nn
     a(i) = a(i) + 1.0_real64
   enddo

   do k = 1, 4
     call sub3 ( a, nn )
   enddo

   return
end subroutine 
!-----------------------------------------------------------------------
!   sub 3
!-----------------------------------------------------------------------
subroutine sub3 ( a, nn )
   use iso_fortran_env, only: real64
   implicit none
   integer,intent(in) :: nn
   real(real64),intent(inout) :: a(nn)
   !*** local variables ***
   integer :: i

   a(1) = sin(a(1))

   do i = 1, nn
     a(i) = a(i) + 1.0_real64
   enddo

   return 
end subroutine 

File: Tuning/sample_code/01_timer/src/c/main.c
/* Copyright 2024 Research Organization for Information Science and Technology */
/*----------------------------------------------------------------------
 *  By-hand timer and Use of gprof 
 *  Author:      Yukihiro Ota (yota@rist.or.jp)
 *--------------------------------------------------------------------*/
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "timer.h"

void sub1 ( double *a, const int nn ) ;
void sub2 ( double *a, const int nn ) ;
void sub3 ( double *a, const int nn ) ;

/*--------------------------------------------------------------------*/
/*  main                                                              */
/*--------------------------------------------------------------------*/
int main ( int argc, char* argv[] )
{
  int nn = 20000 ;
  double a[nn] ;
#if defined(USE_CPU_TIMER)
  double cpu1, cpu2;
#endif
#if defined(USE_ELP_TIMER)
  double elp1, elp2;
#endif

  /* check resolution of timer */
/*#if defined(USE_CPU_TIMER)
  cpu1 = get_cpu_res();
  printf ( "Res. of timer (micro sec) = %17.6e\n", cpu1) ;
#endif
#if defined(USE_ELP_TIMER)
  elp1 = get_elp_res(); 
  printf ( "Res. of timer (micro sec) = %17.6e\n", elp1) ;
#endif*/

  /* set values */
  for ( int i=0; i<nn; ++i ) a[i] = 0.0 ;

  /* routine 1 */
#if defined(USE_CPU_TIMER)
  cpu1 = get_cpu_time () ; // cpu time
#endif
#if defined(USE_ELP_TIMER)
  elp1 = get_elp_time () ; // wall clock
#endif

  for ( int k=0; k<100000; ++k ) sub1 ( a, nn ) ;

#if defined(USE_CPU_TIMER)
  cpu2 = get_cpu_time () ; // cpu time
  printf ( "CPU time (sec)     = %17.6f\n", cpu2 - cpu1 ) ;
#endif
#if defined(USE_ELP_TIMER)
  elp2 = get_elp_time () ; // wall clock
  printf ( "Elapsed time (sec) = %17.6f\n", elp2 - elp1 ) ;
#endif 

  printf ("a[0] = %13.6f\n", a[0] ) ;

  /* routine 2 */
#if defined(USE_CPU_TIMER)
  cpu1 = get_cpu_time () ; // cpu time
#endif
#if defined(USE_ELP_TIMER)
  elp1 = get_elp_time () ; // wall clock
#endif

  for ( int k=0; k<200000; ++k ) sub2 ( a, nn ) ;

#if defined(USE_CPU_TIMER)
  cpu2 = get_cpu_time () ; // cpu time
  printf ( "CPU time (sec)     = %17.6f\n", cpu2 - cpu1 ) ;
#endif
#if defined(USE_ELP_TIMER)
  elp2 = get_elp_time () ; // wall clock
  printf ( "Elapsed time (sec) = %17.6f\n", elp2 - elp1 ) ;
#endif

  printf ("a[0] = %13.6f\n", a[0] ) ;

  /* finalization */
  return EXIT_SUCCESS ;
}
/*--------------------------------------------------------------------*/
/*  sub1                                                              */
/*--------------------------------------------------------------------*/
void sub1 ( double *a, const int nn ) {

  for ( int i=0; i<nn; ++i ) a[i] += 1.0 ;

  for ( int k=0; k<2; ++k ) sub3( a, nn ) ;

}
/*--------------------------------------------------------------------*/
/*  sub2                                                              */
/*--------------------------------------------------------------------*/
void sub2 ( double *a, const int nn ) {

  a[0] = sin(a[0]) ;

  for ( int i=0; i<nn; ++i ) a[i] += 1.0 ;

  for ( int k=0; k<4; ++k ) sub3( a, nn ) ;

}
/*--------------------------------------------------------------------*/
/*  sub3                                                              */
/*--------------------------------------------------------------------*/
void sub3 ( double *a, const int nn ) {

  a[0] = sin(a[0]) ;

  for ( int i=0; i<nn; ++i ) a[i] += 1.0 ;

}
